src/Compilers/VisualBasic/Portable/BoundTree/Statement.vb

```diff
@@ -700,75 +771,30 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-
-        Private ReadOnly Property IVariables As ImmutableArray(Of IVariable) Implements IVariableDeclarationStatement.Variables
-            Get
-                Return Me.LocalDeclarations.As(Of IVariable)()
-            End Get
-        End Property
-
-        Protected Overrides ReadOnly Property IInitialValue As IExpression
-            Get
-                Return Me.Initializer
-            End Get
-        End Property
-
-        Protected Overrides ReadOnly Property IVariable As ILocalSymbol
-            Get
-                ' ZZZ Get clear about what's happening in the VB bound trees. BoundAsNewLocalDeclarations has multiple symbols and
-                ' inherits from BoundLocalDeclarationBase, which occurs multiply in BoundDimStatement.
-                Dim local As BoundLocalDeclaration = Me.LocalDeclarations.FirstOrDefault()
-                Return If(local IsNot Nothing, local.LocalSymbol, Nothing)
-            End Get
-        End Property
-
-        Protected Overrides Function StatementKind() As OperationKind
-            Return OperationKind.VariableDeclarationStatement
-        End Function
-    End Class
-
     Partial Class BoundDimStatement
         Implements IVariableDeclarationStatement
 
+        Private Shared ReadOnly s_variablesMappings As New System.Runtime.CompilerServices.ConditionalWeakTable(Of BoundDimStatement, Object)
+
         Private ReadOnly Property IVariables As ImmutableArray(Of IVariable) Implements IVariableDeclarationStatement.Variables
             Get
-                Return Me.LocalDeclarations.As(Of IVariable)()
+                Dim variables = s_variablesMappings.GetValue(Me, Function(dimStatement)
+                                                                     Dim builder = ArrayBuilder(Of IVariable).GetInstance()
+                                                                     For Each base In dimStatement.LocalDeclarations
+                                                                         If base.Kind = BoundKind.LocalDeclaration Then
+                                                                             Dim declaration = DirectCast(base, BoundLocalDeclaration)
+                                                                             builder.Add(New VariableDeclaration(declaration.LocalSymbol, declaration.InitializerOpt, declaration.Syntax))
+                                                                         ElseIf base.Kind = BoundKind.AsNewLocalDeclarations Then
+                                                                             Dim asNewDeclarations = DirectCast(base, BoundAsNewLocalDeclarations)
+                                                                             For Each asNewDeclaration In asNewDeclarations.LocalDeclarations
+                                                                                 builder.Add(New VariableDeclaration(asNewDeclaration.LocalSymbol, asNewDeclarations.Initializer, asNewDeclaration.Syntax))
```

Comment:
Consider extracting `BoundLocalDeclaration.ToVariable` helper method for here and above.